src/effect/tap_tempo.rs

```diff
@@ -0,0 +1,151 @@
+use crate::{audio::midi::Message, config::NoteOn, effect::Tempo};
+use wmidi::MidiMessage;
+use State::*;
+
+#[derive(Debug)]
+enum State {
+    Start,
+    TappedOnce(u64),
+}
+
+#[derive(Debug)]
+pub struct TapTempo {
+    note_on: NoteOn,
+    state: State,
+}
+
+impl TapTempo {
+    const MAX_DELAY_BETWEEN_TAPS: u64 = 2_000_000; // microseconds
+
+    pub fn new(note_on: NoteOn) -> Self {
+        Self {
+            note_on,
+            state: State::Start,
+        }
+    }
+
+    pub fn handle_messages(&mut self, messages: &[Message]) -> Option<Tempo> {
```

Comment:
`TapTempo` is a state machine which outputs a new `Tempo` whenever it's appropriate.